src/malli/design_draft.clj

```diff
@@ -0,0 +1,96 @@
+(ns malli.design-draft
+  (:import (clojure.lang IPersistentMap)))
+
+
+;;; malli schemas
+
+;; Core schemas are not namespace qualified keywords to make using them user
+;; friendly.
+
+;; Schemas are written using Hiccup style vectors, where order is significant.
+
+;; Every schema definition can have additional options-map after schema name.
+;; Options-map can have arbitrary amount of data, with some keys having special
+;; meaning, like below, which describes a closed map.
+
+[:map {:closed true, :description "My closed map"}]
+
+;; Schema definitions live in registry. Malli provides a core registry for
+;; all the essential schemas. Registry can be extended on the runtime and it's
+;; possible to have as many registries one needs.
+
+;; The actual registry definitions and format is wip, but could be something
+;; like below.
+
+;; Compiles given definition to Schema
+(declare compile)
+(declare Schema)
+
+(def -map (compile {:id :map
+                    :type IPersistentMap
```

Comment:
we should target clojurescript, this is a Clojure type.

Reply:
Good point

Reply:
here are spec-tools types: https://github.com/metosin/spec-tools/blob/master/src/spec_tools/parse.cljc#L66-L86. Should we just use the schema name as a type if it is not defined? e.f. `:map` is a `:map` unless it has a spesidic `:type` overriding it? e.g. having a custom (non-serializable) function/reified schema, one could add a type hint for it. Spec-tools has this, I think it's a one of the good parts.

Also, I guess we want to support all platform-types, so one can say that something is an `Interger` and something is a `Long`. Needed for example in java-interop (schema for a function that calls a `MyJavaLogic(Short s)`.

Reply:
the function Schemas... if we are going to support reified Schemas in schema values, is there a reason not to support functions and coerce them in to Schemas in the `expand` (or whatever) phase? a optional registry from function -> schema would be great?

```clj
;; all data, ok
[:and :int [:> 18]]

:: reified, ok, doesn't de/serialize unless `read` and `write` functions of Schema protocol are implemented, like EDN readers.
[:and (reify Schema (validator [_ value] (number? value))) [:> 18]]

;; explicit coercion of int? -> :int 
[:and int? [:> 18]]

;; fails, no conversion from int? => Schema
(p/schema
  [:and int? [:> 18]])

;; mega-success!
(p/schema
  [:and int? [:> 18]]
  {:registry (merge p/default-registry {int? :int})})
```